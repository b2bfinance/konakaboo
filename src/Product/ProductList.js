import React, { useContext } from "react";
import ProductWrapper from "./ProductWrapper";
import { EmbedContext } from "../EmbedWrapper";

const ProductList = () => {
  const {
    products: [products]
  } = useContext(EmbedContext);

  return products.data.map(product => (
    <ProductWrapper
      key={product.id}
      highlighted={product.highlighted}
      labels={product.labels}
      title={product.title}
      links={product.links}
      brand={product.brand}
      columns={product.columns}
      description={product.description}
      featurePoint={product.feature_point}
      detailed={product.detailed}
      disclaimer={product.disclaimer}
      meta={product.meta}
    />
  ));
};

export default ProductList;
