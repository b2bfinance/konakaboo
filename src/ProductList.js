import { makeStyles, Typography } from "@material-ui/core";
import React from "react";
import { useProductFetcher, useProductState, useConfigState } from "./hooks";
import ProductWrapper from "./ProductWrapper";
import ProductMask from "./ProductMask";

const useStyles = makeStyles(theme => ({
  productListWrapper: {
    opacity: props => (props.loading ? 0.75 : 1)
  },
  productListError: {
    backgroundColor: theme.palette.error.main,
    padding: theme.spacing(2),
    color: theme.palette.error.contrastText
  }
}));

const ProductList = () => {
  const { loading, error, data } = useProductState();
  const { loadingErrorMessage } = useConfigState();

  /**
   * Fetch our products and dispatch them through the product dispatcher which will
   * update our `productState`.
   */
  useProductFetcher();

  const classes = useStyles({
    loading: loading
  });

  if (error) {
    return <Typography className={classes.productListError}>{loadingErrorMessage}</Typography>;
  }

  if (loading && data.length === 0) {
    return (
      <React.Fragment>
        <ProductMask />
        <ProductMask />
        <ProductMask />
        <ProductMask />
        <ProductMask />
      </React.Fragment>
    );
  }

  return (
    <div className={classes.productListWrapper}>
      {data.map((product, i) => (
        <ProductWrapper
          key={product.id || i}
          highlighted={product.highlighted}
          labels={product.labels}
          title={product.title}
          links={product.links}
          brand={product.brand}
          columns={product.columns}
          description={product.description}
          featurePoint={product.feature_point}
          detailed={product.detailed}
          disclaimer={product.disclaimer}
          meta={product.meta}
          product={product}
        />
      ))}
    </div>
  );
};

export default ProductList;
