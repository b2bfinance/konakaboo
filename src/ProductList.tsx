import { Box, Button, Grid, Typography } from "@mui/material";
import React, { useState } from "react";
import { makeStyles } from "tss-react/mui";
import { ProductWrapper } from "./ProductWrapper";
import { Product } from "./ProductTypes";

const useProductListEmptyStyles = makeStyles({
  name: "TabloProductListEmpty",
})((theme) => ({
  root: {
    margin: theme.spacing(10, 0),
    textAlign: "center",
  },
}));

const ProductListEmpty = () => {
  const { classes } = useProductListEmptyStyles();

  return (
    <Grid className={classes.root} container>
      <Grid item xs={12}>
        <Typography variant="h1">No products found</Typography>
      </Grid>
      <Grid item xs={12}>
        <Typography color="textSecondary" paragraph>
          It looks like we couldn't find any products matching your criteria.
        </Typography>
      </Grid>
    </Grid>
  );
};

export type ProductListProps = {
  products: Product[];
  limit: number;
  onMoreDetails: (product: Product) => void;
  onApply: (product: Product) => void;
  cta: string;
  ribbonText: string;
};

const useProductListStyles = makeStyles({
  name: "TabloProductList",
})(() => ({
  root: {},
}));

export const ProductList: React.FC<ProductListProps> = ({
  products,
  limit,
  onMoreDetails,
  onApply,
  cta,
  ribbonText,
}) => {
  const { classes } = useProductListStyles();
  const [productCount, setProductCount] = useState(limit);

  const handleLoadMore = () => {
    setProductCount(productCount + limit);
  };

  if (products.length === 0) {
    return <ProductListEmpty />;
  }

  return (
    <Box className={classes.root}>
      {products.slice(0, productCount).map((product, i) => (
        <ProductWrapper
          key={product.id || i}
          product={product}
          onMoreDetails={onMoreDetails}
          onApply={onApply}
          cta={cta}
          ribbonText={ribbonText}
        />
      ))}
      {products.length > productCount && (
        <Grid container>
          <Grid item>
            <Button
              variant="contained"
              color="secondary"
              onClick={handleLoadMore}
            >
              Load more
            </Button>
          </Grid>
        </Grid>
      )}
    </Box>
  );
};
