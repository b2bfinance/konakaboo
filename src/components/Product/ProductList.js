import React from 'react';
import Product from './Product';
import Row from '../Row';
import Col from '../Col';
import styled, { css } from 'styled-components';
import { queries } from '../../utils/media';

export const StyledProductError = styled.div`
  background-color: #ff9800;
  color: #fff;
  padding: 1rem;
`;

export const ProductMask = styled.div`
  background-color: ${props => props.theme.productMaskBackground};
  height: 35rem;
  margin-bottom: 1.5rem;

  ${queries.desktop`
    height: 9rem;
  `};
`;

export const ProductEmptyContainer = styled(Col)`
  background-color: ${props => props.theme.productEmptyBackground};
  margin: 2rem auto;
  padding: 2rem;
`;

export const ProductsLoadingContainer = styled.div`
  ${props =>
    props.loading &&
    css`
      opacity: 0.75;
    `};
`;

export const ProductError = () => (
  <StyledProductError>
    We had problems retrieving products for you, retry or come back later.
  </StyledProductError>
);

export const ProductLoadingMask = () => (
  <div>
    <ProductMask />
    <ProductMask />
    <ProductMask />
  </div>
);

export const EmptyProducts = () => (
  <Row>
    <ProductEmptyContainer phone="100" desktop="35">
      <p>We don't have any products for you to compare.</p>
    </ProductEmptyContainer>
  </Row>
);

const ProductList = ({ error, hasFetched, isFetching, products, dispatch }) => {
  if (!hasFetched) {
    return <ProductLoadingMask />;
  }

  if (error) {
    return <ProductError />;
  }

  if (products.length === 0) {
    return <EmptyProducts />;
  }

  return (
    <ProductsLoadingContainer loading={isFetching}>
      {products.map((product, i) => (
        <Product key={i} product={product} />
      ))}
    </ProductsLoadingContainer>
  );
};

export default ProductList;
