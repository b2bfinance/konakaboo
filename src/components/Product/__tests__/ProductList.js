import React from 'react';
import { shallow, render } from 'enzyme';
import ProductList, {
  StyledProductError,
  ProductMask,
  ProductEmptyContainer,
  ProductsLoadingContainer,
  ProductError,
  ProductLoadingMask,
  EmptyProducts
} from '../ProductList';

test('StyledProductError renders correctly', () => {
  expect(render(<StyledProductError />)).toMatchSnapshot();
});

test('ProductMask renders correctly', () => {
  expect(render(<ProductMask theme={stubData.theme} />)).toMatchSnapshot();
});

test('ProductEmptyContainer renders correctly', () => {
  expect(
    render(<ProductEmptyContainer theme={stubData.theme} />)
  ).toMatchSnapshot();
});

test('ProductsLoadingContainer renders correctly', () => {
  expect(render(<ProductsLoadingContainer />)).toMatchSnapshot();
});

test('ProductsLoadingContainer while loading renders correctly', () => {
  expect(render(<ProductsLoadingContainer loading />)).toMatchSnapshot();
});

test('ProductError renders correctly', () => {
  expect(render(<ProductError />)).toMatchSnapshot();
});

test('ProductLoadingMask renders correctly', () => {
  expect(render(<ProductLoadingMask />)).toMatchSnapshot();
});

test('EmptyProducts renders correctly', () => {
  expect(render(<EmptyProducts />)).toMatchSnapshot();
});

test('ProductList when not fetched renders correctly', () => {
  expect(render(<ProductList hasFetched={false} />)).toMatchSnapshot();
});

test('ProductList with an error renders correctly', () => {
  expect(
    render(<ProductList hasFetched={true} error={true} />)
  ).toMatchSnapshot();
});

test('ProductList with no products renders correctly', () => {
  expect(
    render(<ProductList hasFetched={true} products={[]} />)
  ).toMatchSnapshot();
});

test('ProductList with products renders correctly', () => {
  expect(
    shallow(
      <ProductList
        hasFetched={true}
        products={stubData.products.response.data}
      />
    )
  ).toMatchSnapshot();
});
