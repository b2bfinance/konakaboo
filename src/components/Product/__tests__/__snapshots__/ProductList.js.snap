// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmptyProducts renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
}

.c1 {
  width: 100%;
  min-height: 1px;
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  max-width: 100%;
  margin: 2rem auto;
  padding: 2rem;
}

@media (min-width:62em) {
  .c1 {
    -webkit-flex: 0 0 35%;
    -ms-flex: 0 0 35%;
    flex: 0 0 35%;
    max-width: 35%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <p>
      We don't have any products for you to compare.
    </p>
  </div>
</div>
`;

exports[`ProductEmptyContainer renders correctly 1`] = `
.c0 {
  width: 100%;
  min-height: 1px;
  background-color: #eceff1;
  margin: 2rem auto;
  padding: 2rem;
}

<div
  class="c0"
>
  
</div>
`;

exports[`ProductError renders correctly 1`] = `
.c0 {
  background-color: #ff9800;
  color: #fff;
  padding: 1rem;
}

<div
  class="c0"
>
  We had problems retrieving products for you, retry or come back later.
</div>
`;

exports[`ProductList when not fetched renders correctly 1`] = `
.c0 {
  height: 35rem;
  margin-bottom: 1.5rem;
}

@media (min-width:62em) {
  .c0 {
    height: 9rem;
  }
}

<div>
  <div
    class="c0"
  >
    
  </div>
  <div
    class="c0"
  >
    
  </div>
  <div
    class="c0"
  >
    
  </div>
</div>
`;

exports[`ProductList with an error renders correctly 1`] = `
.c0 {
  background-color: #ff9800;
  color: #fff;
  padding: 1rem;
}

<div
  class="c0"
>
  We had problems retrieving products for you, retry or come back later.
</div>
`;

exports[`ProductList with no products renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  width: 100%;
}

.c1 {
  width: 100%;
  min-height: 1px;
  -webkit-flex: 0 0 100%;
  -ms-flex: 0 0 100%;
  flex: 0 0 100%;
  max-width: 100%;
  margin: 2rem auto;
  padding: 2rem;
}

@media (min-width:62em) {
  .c1 {
    -webkit-flex: 0 0 35%;
    -ms-flex: 0 0 35%;
    flex: 0 0 35%;
    max-width: 35%;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <p>
      We don't have any products for you to compare.
    </p>
  </div>
</div>
`;

exports[`ProductList with products renders correctly 1`] = `
<styled.div>
  <Product
    key="0"
    product={
      Object {
        "brand": "Landbay",
        "columns": Array [
          Object {
            "label": "Max LTV",
            "value": "85%",
          },
          Object {
            "label": "Initial Rate",
            "value": "1.49%",
          },
          Object {
            "label": "Subsequent Rate",
            "value": "2.75%",
          },
          Object {
            "label": "Fees",
            "value": "£1,495",
          },
        ],
        "description": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean bibendum facilisis condimentum. Donec laoreet porttitor condimentum. Nulla luctus efficitur magna, eget rhoncus metus placerat sed.",
        "highlighted": true,
        "highlighted_points": Array [
          "Online Application",
          "Mobile app",
          "Customer service winner",
        ],
        "labels": Array [
          "broker only",
          "consumer favourite",
        ],
        "links": Object {
          "apply": "https://b2bfinance.com/enquire/buy-to-let",
          "logo": "https://b2bfinance.com/images/partners/Landbay.png",
        },
        "meta": Object {
          "confirm": Object {
            "description": "B2Bfinance.com are registered intermediaries who will quote, package and submit your lending request to Natwest. There is no charge for this service as we are paid a small commission by Natwest to cover our time.",
            "heading": "Application via Brokers only",
          },
        },
        "technical_points": Array [
          "1.5% Early Repayment",
          "6 months cool off peroid",
          "No monthly fee for existing customers",
        ],
        "title": "Landbay - Buy To Let Wizards",
      }
    }
  />
  <Product
    key="1"
    product={
      Object {
        "brand": "Natwest",
        "columns": Array [
          Object {
            "label": "Max LTV",
            "value": "85%",
          },
          Object {
            "label": "Initial Rate",
            "value": "1.52%",
          },
          Object {
            "label": "Subsequent Rate",
            "value": "3.25%",
          },
          Object {
            "label": "Fees",
            "value": "£995",
          },
        ],
        "description": "",
        "highlighted": false,
        "highlighted_points": Array [],
        "labels": Array [],
        "links": Object {
          "apply": "https://b2bfinance.com/enquire/buy-to-let",
          "logo": "https://b2bfinance.com/images/partners/NatWest.png",
        },
        "meta": Object {
          "confirm": Object {
            "description": "B2Bfinance.com are registered intermediaries who will quote, package and submit your lending request to Natwest. There is no charge for this service as we are paid a small commission by Natwest to cover our time.",
            "heading": "Application via Brokers only",
          },
        },
        "technical_points": Array [],
        "title": "Natwest Low Low Rate",
      }
    }
  />
  <Product
    key="2"
    product={
      Object {
        "brand": "Barclays",
        "columns": Array [
          Object {
            "label": "Max LTV",
            "value": "80%",
          },
          Object {
            "label": "Initial Rate",
            "value": "1.75%",
          },
          Object {
            "label": "Subsequent Rate",
            "value": "3%",
          },
          Object {
            "label": "Fees",
            "value": "£995",
          },
        ],
        "description": "Whether your landlord is a first time buyer, a first time landlord or has a large portfolio, we have a range of products to help. Our range of mortgages also includes products for Limited Companies and HMOs.",
        "highlighted": false,
        "highlighted_points": Array [
          "Complete online application",
          "Mobile app",
          "Customer service winner",
        ],
        "labels": Array [
          "broker only",
        ],
        "links": Object {
          "apply": "https://b2bfinance.com/enquire/buy-to-let",
          "logo": "https://b2bfinance.com/images/partners/Precise.png",
        },
        "meta": Object {},
        "technical_points": Array [
          "Rental cover - ICR: based on individual circumstances",
          "5 year fixed rate products that help maximise the loan size available",
          "5 year fixed with 4% cashback available on purchases",
        ],
        "title": "Precise Oh So Precise Total Winners Deal",
      }
    }
  />
</styled.div>
`;

exports[`ProductLoadingMask renders correctly 1`] = `
.c0 {
  height: 35rem;
  margin-bottom: 1.5rem;
}

@media (min-width:62em) {
  .c0 {
    height: 9rem;
  }
}

<div>
  <div
    class="c0"
  >
    
  </div>
  <div
    class="c0"
  >
    
  </div>
  <div
    class="c0"
  >
    
  </div>
</div>
`;

exports[`ProductMask renders correctly 1`] = `
.c0 {
  background-color: #f5f6f7;
  height: 35rem;
  margin-bottom: 1.5rem;
}

@media (min-width:62em) {
  .c0 {
    height: 9rem;
  }
}

<div
  class="c0"
>
  
</div>
`;

exports[`ProductsLoadingContainer renders correctly 1`] = `
<div
  class=""
>
  
</div>
`;

exports[`ProductsLoadingContainer while loading renders correctly 1`] = `
.c0 {
  opacity: 0.75;
}

<div
  class="c0"
>
  
</div>
`;

exports[`StyledProductError renders correctly 1`] = `
.c0 {
  background-color: #ff9800;
  color: #fff;
  padding: 1rem;
}

<div
  class="c0"
>
  
</div>
`;
